### ISSUE TITLE
how to control simple_switch_grpc?

### ISSUE THREAD

**Issue by MaoJianwei** [p4lang/PI#330]
Hi,

Can PI be used to control simple_switch_grpc target? I don't find instructions at Readme.

Thanks!

**Comment by antoninbas** [comment]
You are supposed to use the P4Runtime protobuf interface to control simple_switch_grpc: https://github.com/p4lang/PI/blob/master/proto/p4/p4runtime.proto. I recommend that you first get familiar with protobuf and gRPC. You can also take a look at the example application in the bmv2 repo: https://github.com/p4lang/behavioral-model/blob/master/targets/simple_switch_grpc/tests/example.cpp. Finally, we are currently working on some documentaion for P4Runtime (https://github.com/p4lang/PI/tree/master/proto/docs) but it is far from being complete.

**Comment by coolvikas** [comment]
set_queue_rate and set_queue_depth are not yet implemented in p4runtime.proto Is there any way I can configure simple_switch_grpc with them?

**Comment by MaoJianwei** [comment]
Hi antoninbas @antoninbas ,

I understand protobuf and gRPC, 
so Is PI a tool just for thrift? but I see '--with-proto' in the README.

I can use ONOS controller to control simple_switch_grpc now, and I want to check simple_switch_grpc  directly by another way, (like ovs-ofctl for ovs, Simple_Switch_CLI for simple_switch).

So Is there any way to check the status of simple_switch_grpc? using PI or other tools?


Thanks!

**Comment by antoninbas** [comment]
I'm not sure what you mean by "PI is just a tool for thrift".

The PI repository includes an implementation of a gRPC server which supports the P4Runtime service. In order to build and install the gRPC server, you need to `configure` the PI code with the `--with-proto` flag.

The P4Runtime service calls into the PI C library, which is implemented in this repo (https://github.com/p4lang/behavioral-model/tree/master/PI). In order to build this code, you need to `configure` bmv2 with `--with-pi`. There is an alternative implementation of the PI library which uses Thrift (https://github.com/p4lang/PI/tree/master/targets/bmv2). That implementation has the advantage that it supports multiple devices with the same P4Runtime gRPC server, but I would not recommend using it unless you are already familiar with simple_switch_grpc.

So in a nutshell:
 - build and install PI: `./autogen.sh && ./configure --with-proto --without-internal-rpc --without-cli --without-bmv2 && make && sudo make install && sudo ldconfig`
 - build bmv2: `./autogen.sh && ./configure --with-pi && make`
 - build simple_switch_grpc: `cd targets/simple_switch_grpc && ./autogen.sh && ./configure && make`

If you need a separate way of inspecting the P4 dataplane, you will need to enable the Thrift server on simple_switch_grpc (https://github.com/p4lang/behavioral-model/tree/master/targets/simple_switch_grpc#enabling-the-thrift-server) and use the simple_switch_CLI. You will then essentially have 2 RPC servers running in your simple_switch_grpc process: the P4Runtime gRPC server and the Thrift RPC server.

**Comment by antoninbas** [comment]
I added more thorough build instructions to the simple_switch_grpc README as well as a FAQ: https://github.com/p4lang/behavioral-model/tree/master/targets/simple_switch_grpc

**Comment by MaoJianwei** [comment]
Thanks a lot, and I also think using thrift to inspect simple_switch_grpc is not a good idea.

I think the newest README and FAQ can't answer my question yet. And I can make my question more clear:

What I want is not a P4Runtime Server for multiple devices or a Thrift Client Tool for a simple_switch_grpc device.

I want a **P4Runtime Client** to inspect a simple_switch_grpc device ( **P4Runtime Server** ).

For example, analogously:
simple_switch_CLI is **Thrift Client** for simple_switch device (**Thrift Server**)
ovs-vsctl is **OVSDB Client** for Openvswitch device(ovs-vswitched is **OVSDB Server**)


Thanks!

**Comment by antoninbas** [comment]
There is no such P4Runtime CLI-like client for now. Maybe one will be developed as part of the Stratum project (https://stratumproject.org/). Of course you can always try using the generic gRPC cli tool (https://github.com/grpc/grpc/blob/master/doc/command_line_tool.md) but I personally do not find it very convenient. You can also develop one yourself in the language of your choice (and ideally contribute it to the community).

The absence of a CLI is why we provided to enable a Thrift server for simple_switch_grpc in the first place, so that simple_switch_CLI can be used for inspecting dataplane tables.
